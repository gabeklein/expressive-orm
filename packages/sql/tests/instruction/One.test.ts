import { Num, One, Query, Str, Type } from '../..';

class A extends Type {
  b = One(B);
  value = Str();
}

class B extends Type {
  c = One(C);
}

class C extends Type {
  value = Num();
  label = Str();
}

it("will query via direct selection", () => {
  const query = new Query(where => {
    const a = where(A);

    return where.selects(a.value);
  });

  expect(query).toMatchSnapshot();
})

it("will query via select function", () => {
  const query = new Query(where => {
    const a = where(A);

    return where.selects(() => a.value);
  });

  expect(query).toMatchSnapshot();
})

it("will select via an object", () => {
  const query = new Query(where => {
    const a = where(A);

    return where.selects({
      aValue: a.value,
      cValue: a.b.c.value
    })
  });

  expect(query).toMatchSnapshot();
})

it("will query nested relationships", () => {
  // TODO: fix types
  const query = new Query(where => {
    const a = where(A);

    where(a.b.c.value).is(100);
    
    return where.selects(a.b.c.label);
  })

  expect(query).toMatchSnapshot();
})