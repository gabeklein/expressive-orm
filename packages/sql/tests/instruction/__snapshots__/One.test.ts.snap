// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`will query nested relationships 1`] = `
	select
	  \`C\`.\`label\` as \`label\`
	from
	  \`A\`
	  left join \`B\` on \`B\`.\`id\` = \`A\`.\`bId\`
	  left join \`C\` on \`C\`.\`id\` = \`B\`.\`cId\`
	where
	  \`C\`.\`value\` = 100
`;

exports[`will query via select function 1`] = `
	select
	  \`A\`.\`value\` as 1
	from
	  \`A\`
	  left join \`B\` on \`B\`.\`id\` = \`A\`.\`bId\`
	  left join \`C\` on \`C\`.\`id\` = \`B\`.\`cId\`
`;

exports[`will select via an object 1`] = `
	select
	  \`A\`.\`value\` as \`aValue\`,
	  \`C\`.\`value\` as \`cValue\`
	from
	  \`A\`
	  left join \`B\` on \`B\`.\`id\` = \`A\`.\`bId\`
	  left join \`C\` on \`C\`.\`id\` = \`B\`.\`cId\`
`;
